<a href="https://colab.research.google.com/github/wahyuimammuhroni-jpg/ecommerce-order-supply-chain-analysis/blob/main/notebooks/ecommerce_order_supply_chain_analytics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Analisis Dataset "E-Commerce Order & Supply Chain"

Dataset ini berisi data transaksi e-commerce yang mencakup informasi mengenai pesanan, produk, pelanggan, pembayaran, seller, serta proses pengiriman. Analisis ini bertujuan untuk memahami performa bisnis dan operasional e-commerce, khususnya dalam melihat pola penjualan, performa produk dan seller, metode pembayaran, serta efektivitas proses pengiriman kepada pelanggan.

Dataset terdiri dari beberapa tabel yang saling berhubungan melalui `order_id`, `customer_id`, dan `product_id`.

Beberapa dataset dan kolom penting yang digunakan dalam analisis ini meliputi:

- **Orders** : Informasi utama mengenai pesanan dan proses pengiriman.
  - `order_id` : ID unik setiap pesanan.
  - `customer_id` : ID pelanggan yang melakukan pesanan.
  - `order_status` : Status pesanan.
  - `order_purchase_timestamp` : Waktu ketika pesanan dibuat.
  - `order_delivered_customer_date` : Waktu ketika pesanan diterima pelanggan.
  - `order_estimated_delivery_date` : Estimasi waktu pengiriman.

- **Order Items** : Informasi produk dan seller pada setiap transaksi.
  - `order_id` : ID pesanan.
  - `product_id` : ID produk.
  - `seller_id` : ID seller.
  - `price` : Harga produk.
  - `shipping_charges` : Biaya pengiriman.

- **Products** : Informasi mengenai produk yang dijual.
  - `product_id` : ID unik produk.
  - `product_category_name` : Kategori produk.
  - Informasi dimensi dan berat produk.

- **Customers** : Informasi lokasi pelanggan.
  - `customer_id` : ID pelanggan.
  - `customer_city` : Kota pelanggan.
  - `customer_state` : State atau wilayah pelanggan.

- **Payments** : Informasi mengenai metode pembayaran transaksi.
  - `order_id` : ID pesanan.
  - `payment_type` : Metode pembayaran.
  - `payment_installments` : Jumlah cicilan.
  - `payment_value` : Nilai pembayaran.

Dalam analisis ini, kita akan mengeksplorasi hubungan antar tabel untuk mengidentifikasi pola penjualan, performa kategori produk dan seller, tren transaksi, performa pengiriman, serta perbedaan performa operasional berdasarkan wilayah pelanggan.

Proses analisis akan menggunakan **Python melalui Google Colab** untuk data understanding, data cleaning, feature engineering, exploratory data analysis, dan visualisasi. Selanjutnya, **Google BigQuery dan SQL** akan digunakan untuk melakukan pengolahan data relasional, agregasi, serta analisis bisnis yang lebih mendalam.

## Import Library & Dataset

In [3]:
# Import Library
import pandas as pd
import numpy as np
from IPython.display import display

orders = pd.read_csv("/content/df_Orders.csv")
order_items = pd.read_csv("/content/df_OrderItems.csv")
products = pd.read_csv("/content/df_Products.csv")
customers = pd.read_csv("/content/df_Customers.csv")
payments = pd.read_csv("/content/df_Payments.csv")

print("Dataset berhasil dibaca.")


Dataset berhasil dibaca.


## Data Understanding  

Pada tahap ini berfokus pada persiapan kumpulan data sebelum melakukan *Exploratory Data Analysis* (EDA). Proses ini mencakup pemeriksaan dan identifikasi missing values, identifikasi data duplikat, validasi tipe data dan kategori, ringkasan statistik deskriptif serta memastikan kumpulan data konsisten dan siap untuk analisis lebih lanjut.

In [4]:
# Check All Dataset
datasets = {
    "Orders": orders,
    "Order Items": order_items,
    "Products": products,
    "Customers": customers,
    "Payments": payments
}

for name, df in datasets.items():
    print(f"\n{name}")
    display(df.head())
    print(f"{'='*100}")


Orders


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_timestamp,order_estimated_delivery_date
0,Axfy13Hk4PIk,hCT0x9JiGXBQ,delivered,2017-10-22 18:57:54,2017-10-22 19:14:13,2017-10-26 22:19:52,2017-11-09
1,v6px92oS8cLG,PxA7fv9spyhx,delivered,2018-06-20 21:40:31,2018-06-20 22:20:20,2018-07-03 22:51:22,2018-07-24
2,Ulpf9skrhjfm,g3nXeJkGI0Qw,delivered,2018-02-16 16:19:31,2018-02-17 16:15:35,2018-02-27 01:29:50,2018-03-08
3,bwJVWupf2keN,EOEsCQ6QlpIg,delivered,2018-08-18 18:04:29,2018-08-18 18:15:16,2018-08-27 20:03:51,2018-09-19
4,Dd0QnrMk9Cj5,mVz5LO2Vd6cL,delivered,2017-12-22 16:44:04,2017-12-22 17:31:31,2018-01-05 19:22:49,2018-01-18



Order Items


,order_id,product_id,seller_id,price,shipping_charges
0,Axfy13Hk4PIk,90K0C1fIyQUf,ZWM05J9LcBSF,223.51,84.65
1,v6px92oS8cLG,qejhpMGGVcsl,IjlpYfhUbRQs,170.80,23.79
2,Ulpf9skrhjfm,qUS5d2pEAyxJ,77p2EYxcM9MD,64.40,17.38
3,bwJVWupf2keN,639iGvMyv0De,jWzS0ayv9TGf,264.50,30.72
4,Dd0QnrMk9Cj5,1lycYGcsic2F,l1pYW6GBnPMr,779.90,30.66



Products


,product_id,product_category_name,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,90K0C1fIyQUf,toys,491.0,19.0,12.0,16.0
1,qejhpMGGVcsl,watches_gifts,440.0,18.0,14.0,17.0
2,qUS5d2pEAyxJ,costruction_tools_garden,2200.0,16.0,16.0,16.0
3,639iGvMyv0De,toys,1450.0,68.0,3.0,48.0
4,1lycYGcsic2F,toys,300.0,17.0,4.0,12.0



Customers


,customer_id,customer_zip_code_prefix,customer_city,customer_state
0,hCT0x9JiGXBQ,58125,varzea paulista,SP
1,PxA7fv9spyhx,3112,armacao dos buzios,RJ
2,g3nXeJkGI0Qw,4119,jandira,SP
3,EOEsCQ6QlpIg,18212,uberlandia,MG
4,mVz5LO2Vd6cL,88868,ilhabela,SP



Payments


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,Axfy13Hk4PIk,1,credit_card,1,259.14
1,v6px92oS8cLG,1,credit_card,8,382.39
2,Ulpf9skrhjfm,1,credit_card,4,249.25
3,bwJVWupf2keN,1,credit_card,2,27.79
4,Dd0QnrMk9Cj5,1,credit_card,1,76.15


In [5]:
# Display Rows & Columns in Dataset
for name, df in datasets.items():
    print(f"{name}: {df.shape[0]} Rows dan {df.shape[1]} Column")

Orders: 89316 Rows dan 7 Column
Order Items: 89316 Rows dan 5 Column
Products: 89316 Rows dan 6 Column
Customers: 89316 Rows dan 4 Column
Payments: 89316 Rows dan 5 Column


In [6]:
# Display Information on Data Structures and Data Types
for name, df in datasets.items():
    print(f"\n{'='*50}")
    print(f"Dataset: {name}")
    print(f"{'='*50}")
    df.info()


Dataset: Orders
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89316 entries, 0 to 89315
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       89316 non-null  object
 1   customer_id                    89316 non-null  object
 2   order_status                   89316 non-null  object
 3   order_purchase_timestamp       89316 non-null  object
 4   order_approved_at              89307 non-null  object
 5   order_delivered_timestamp      87427 non-null  object
 6   order_estimated_delivery_date  89316 non-null  object
dtypes: object(7)
memory usage: 4.8+ MB

Dataset: Order Items
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89316 entries, 0 to 89315
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          89316 non-null  object 
 1   product_id        89316 non-

In [7]:
# Check Missing Values
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.isnull().sum())
    print(f"{'='*50}")


Orders
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   9
order_delivered_timestamp        1889
order_estimated_delivery_date       0
dtype: int64

Order Items
order_id            0
product_id          0
seller_id           0
price               0
shipping_charges    0
dtype: int64

Products
product_id                 0
product_category_name    308
product_weight_g          15
product_length_cm         15
product_height_cm         15
product_width_cm          15
dtype: int64

Customers
customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Payments
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


Ditemukan missing values pada dataset `Orders` dan `Products`.

Pada dataset `Orders`, terdapat 9 missing values pada `order_approved_at` dan 1.889 missing values pada `order_delivered_timestamp`.

Pada dataset `Products`, terdapat 308 missing values pada `product_category_name`, serta masing-masing 15 missing values pada informasi berat dan dimensi produk.

Sementara itu, dataset `Order Items`, `Customers`, dan `Payments` tidak memiliki missing values.

Missing values tersebut akan dianalisis lebih lanjut pada tahap Data Preparation untuk menentukan penanganan yang sesuai.

In [8]:
# Check Duplicate Data
for name, df in datasets.items():
    print(f"{name}: {df.duplicated().sum()} duplicate data")

Orders: 0 duplicate data
Order Items: 0 duplicate data
Products: 61865 duplicate data
Customers: 0 duplicate data
Payments: 0 duplicate data


Terdapat data duplicate pada `Products` memiliki **61.865** baris duplikat. Jumlah tersebut cukup besar sehingga perlu diperiksa lebih lanjut untuk memastikan apakah duplikasi terjadi pada keseluruhan baris atau karena satu `product_id` muncul berulang kali.

In [9]:
# Check Unique Values
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.nunique())
    print(f"{'='*50}")


Orders
order_id                         89316
customer_id                      89316
order_status                         7
order_purchase_timestamp         72775
order_approved_at                68215
order_delivered_timestamp        71143
order_estimated_delivery_date      444
dtype: int64

Order Items
order_id            89316
product_id          27451
seller_id            2929
price                5235
shipping_charges     6327
dtype: int64

Products
product_id               27451
product_category_name       70
product_weight_g          2020
product_length_cm           99
product_height_cm          101
product_width_cm            93
dtype: int64

Customers
customer_id                 89316
customer_zip_code_prefix    13930
customer_city                3735
customer_state                 27
dtype: int64

Payments
order_id                89316
payment_sequential         28
payment_type                4
payment_installments       24
payment_value           25520
dtype: int64


Hasil check unique values menunjukkan beberapa karakteristik penting pada dataset:

- `Orders` memiliki 89.316 `order_id` unik, menunjukkan bahwa setiap baris merepresentasikan satu order.
- `Order Items` memiliki 89.316 `order_id` unik, tetapi hanya 27.451 `product_id` unik dan 2.929 `seller_id` unik.
- `Products` memiliki 27.451 `product_id` unik dari total 89.316 baris. Hal ini mengindikasikan adanya pengulangan informasi produk dan menjelaskan tingginya jumlah data duplikat pada dataset tersebut.
- `Customers` memiliki 89.316 `customer_id` unik yang tersebar pada 3.735 kota dan 27 state.
- `Payments` memiliki 89.316 `order_id` unik dan 4 jenis metode pembayaran.

In [10]:
# Check Descriptive Statistics
for name, df in datasets.items():
    print(f"\n{'='*50}")
    print(f"Dataset: {name}")
    print(f"{'='*50}")
    display(df.describe())


Dataset: Orders


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_timestamp,order_estimated_delivery_date
count,89316,89316,89316,89316,89307,87427,89316
unique,89316,89316,7,72775,68215,71143,444
top,Xo1fqyXOC6CR,rhevVNXYJnEs,delivered,2017-08-08 20:26:31,2017-08-08 20:43:31,2017-08-14 12:46:18,2017-12-20
freq,1,1,87428,87,87,87,500



Dataset: Order Items


,price,shipping_charges
count,89316.000000,89316.000000
mean,340.900543,44.283210
std,557.459897,37.672491
min,0.850000,0.000000
25%,59.650000,20.110000
50%,136.900000,35.055000
75%,399.200000,57.190000
max,6735.000000,409.680000



Dataset: Products


,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,89301.000000,89301.000000,89301.000000,89301.00000
mean,2087.068129,30.220658,16.559423,23.03421
std,3747.039215,16.110199,13.388514,11.68828
min,0.000000,7.000000,2.000000,6.00000
25%,300.000000,18.000000,8.000000,15.00000
50%,700.000000,25.000000,13.000000,20.00000
75%,1800.000000,38.000000,20.000000,30.00000
max,40425.000000,105.000000,105.000000,118.00000



Dataset: Customers


,customer_zip_code_prefix
count,89316.000000
mean,40499.471080
std,31194.386361
min,1003.000000
25%,9785.000000
50%,35480.000000
75%,70040.000000
max,99990.000000



Dataset: Payments


,payment_sequential,payment_installments,payment_value
count,89316.000000,89316.000000,89316.000000
mean,1.099646,2.965717,268.657190
std,0.772658,2.796406,344.409566
min,1.000000,0.000000,0.000000
25%,1.000000,1.000000,84.340000
50%,1.000000,2.000000,171.860000
75%,1.000000,4.000000,313.530000
max,29.000000,24.000000,7274.880000


Deskriptif statistik menunjukkan rentang dan distribusi nilai pada setiap kolom numerik. Beberapa variabel seperti `price`, `shipping_charges`, dimensi produk, berat produk, `payment_installments`, dan `payment_value` akan diperiksa lebih lanjut untuk mengidentifikasi outlier pada dataset

Sementara itu, `customer_zip_code_prefix` tidak digunakan sebagai variabel numerik dalam analisis karena merupakan kode identifikasi wilayah.

In [11]:
# Check Distribution Categorical Data
print("Distribution Order Status:")
display(orders["order_status"].value_counts())

print("\nDistribution Payment Type:")
display(payments["payment_type"].value_counts())

print("\nDistribution Customer State:")
display(customers["customer_state"].value_counts())

print("\nTop 10 Product Category Name:")
display(
    products["product_category_name"]
    .value_counts()
    .head(10)
)

Distribution Order Status:


,count
order_status,
delivered,87428
shipped,936
canceled,409
processing,273
invoiced,266
unavailable,2
approved,2



Distribution Payment Type:


,count
payment_type,
credit_card,65814
wallet,17302
voucher,4911
debit_card,1289



Distribution Customer State:


,count
customer_state,
SP,37879
RJ,11578
MG,10334
RS,4921
PR,4523
SC,3213
BA,3088
GO,1868
ES,1798



Top 10 Product Category Name:


,count
product_category_name,
toys,67027
health_beauty,2351
bed_bath_table,2146
sports_leisure,1837
furniture_decor,1760
computers_accessories,1715
housewares,1340
watches_gifts,1196
telephony,912


Hasil yang ditemukan dari dateset untuk Distribution Categorical Data, yaitu.

`order_status`: Mayoritas pesanan memiliki status `delivered`, yaitu 87.428 order. Hal ini menunjukkan bahwa sebagian besar transaksi dalam dataset telah menyelesaikan proses pengiriman, sementara hanya sebagian kecil order yang masih berada pada tahap shipped, processing, invoiced, canceled, unavailable, atau approved.

`payment_type`: Metode pembayaran paling banyak digunakan adalah `credit_card`, diikuti oleh `wallet`, `voucher`, dan `debit_card`. Dominasi penggunaan credit card menunjukkan bahwa metode pembayaran tersebut merupakan metode utama dalam transaksi pada dataset ini.

`customer_state`: Customer tersebar di 27 state, dengan jumlah customer terbesar berasal dari `SP`, diikuti oleh `RJ` dan `MG`. Distribusi ini menunjukkan bahwa transaksi cukup terkonsentrasi pada beberapa wilayah tertentu.

`product_category_name`: Kategori `toys` muncul jauh lebih dominan dibandingkan kategori lainnya pada dataset `Products`. Namun, karena dataset `Products` masih memiliki sejumlah besar data duplikat, distribusi kategori ini belum dapat langsung diinterpretasikan sebagai performa penjualan. Analisis kategori produk akan dilakukan kembali setelah proses data cleaning.

## Data Preparation

Pada tahap ini dilakukan proses persiapan dan pembersihan data agar dataset siap digunakan untuk analisis lebih lanjut. Proses ini mencakup penyesuaian tipe data, penanganan duplikasi dan missing values, validasi data, serta pembuatan variabel baru yang dibutuhkan dalam analisis.

In [12]:
# Create Copy Dataset
orders_clean = orders.copy()
order_items_clean = order_items.copy()
products_clean = products.copy()
customers_clean = customers.copy()
payments_clean = payments.copy()

In [13]:
# Changing Data Type of Date Column
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_timestamp",
    "order_estimated_delivery_date"
]

for column in date_columns:
    orders_clean[column] = pd.to_datetime(
        orders_clean[column],
        errors="coerce"
    )

In [14]:
# Date Column Data Type Validation
orders_clean[date_columns].dtypes

,0
order_purchase_timestamp,datetime64[ns]
order_approved_at,datetime64[ns]
order_delivered_timestamp,datetime64[ns]
order_estimated_delivery_date,datetime64[ns]


In [15]:
# Handling Duplicate in Product
product_check = (
    products_clean
    .groupby("product_id")
    .nunique(dropna=False)
)

conflicting_products = product_check[
    (product_check > 1).any(axis=1)
]

print("Jumlah product_id dengan informasi yang berbeda:",
      len(conflicting_products))

Jumlah product_id dengan informasi yang berbeda: 0


In [16]:
# Drop Duplicate in Product
products_clean = products_clean.drop_duplicates(
    subset="product_id"
).reset_index(drop=True)

In [17]:
# Check Validation
print("Jumlah baris setelah cleaning:", len(products_clean))
print("Jumlah product_id unique:", products_clean["product_id"].nunique())
print("Jumlah duplicate product_id:",
      products_clean["product_id"].duplicated().sum())

Jumlah baris setelah cleaning: 27451
Jumlah product_id unique: 27451
Jumlah duplicate product_id: 0


Hasil menunjukkan bahwa tidak terdapat `product_id` yang memiliki informasi produk berbeda. Dengan demikian, pengulangan pada dataset `Products` merupakan data duplikat yang konsisten.

Data duplikat kemudian dihapus berdasarkan `product_id`, sehingga jumlah data berkurang dari 89.316 menjadi 27.451 baris dan setiap `product_id` hanya memiliki satu record unik.

Proses ini penting untuk mencegah terjadinya penggandaan nilai ketika dataset `Products` digabungkan dengan dataset transaksi pada tahap analisis.

In [18]:
# Handling Missing Values
for name, df in {
    "Orders": orders_clean,
    "Order Items": order_items_clean,
    "Products": products_clean,
    "Customers": customers_clean,
    "Payments": payments_clean
}.items():
    print(f"\n{name}")
    print(df.isnull().sum())
    print(f"{'='*50}")


Orders
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   9
order_delivered_timestamp        1889
order_estimated_delivery_date       0
dtype: int64

Order Items
order_id            0
product_id          0
seller_id           0
price               0
shipping_charges    0
dtype: int64

Products
product_id                 0
product_category_name    141
product_weight_g           2
product_length_cm          2
product_height_cm          2
product_width_cm           2
dtype: int64

Customers
customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Payments
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


In [19]:
# Check Missing Values in order_delivered_timestamp by Categorical order_status
pd.crosstab(
    orders_clean["order_status"],
    orders_clean["order_delivered_timestamp"].isna()
)

order_delivered_timestamp,False,True
order_status,,
approved,0,2
canceled,5,404
delivered,87422,6
invoiced,0,266
processing,0,273
shipped,0,936
unavailable,0,2


Sebagian besar missing values pada `order_delivered_timestamp` berasal dari pesanan yang belum mencapai status delivered atau telah dibatalkan. Oleh karena itu, missing values tersebut dianggap sesuai dengan kondisi transaksi dan tetap dipertahankan sebagai `NaT`.

Namun, terdapat 6 pesanan dengan status `delivered` yang tidak memiliki `order_delivered_timestamp`. Data tersebut akan tetap dipertahankan, tetapi tidak digunakan dalam analisis durasi pengiriman karena tidak memiliki informasi waktu pengiriman yang lengkap.

In [20]:
# Check Missing Values in order_approved_at by Categorical "Delivered"
orders_clean.loc[
    orders_clean["order_approved_at"].isna(),
    "order_status"
].value_counts()

,count
order_status,
delivered,9


Terdapat 9 pesanan berstatus `delivered` yang tidak memiliki informasi `order_approved_at`. Karena tidak tersedia informasi yang dapat digunakan untuk menentukan waktu approval yang sebenarnya, missing values tersebut tetap dipertahankan dan akan dikecualikan apabila dilakukan analisis terkait waktu approval.

In [21]:
# Handling Missing Values in column product_category_name by entering "Unknown"
products_clean["product_category_name"] = (
    products_clean["product_category_name"]
    .fillna("Unknown")
)

In [22]:
# Check Validation
products_clean["product_category_name"].isnull().sum()

np.int64(0)

Missing values pada `product_category_name` diisi dengan kategori `Unknown`. Hal tersebut digunakan agar produk tetap dapat dipertahankan dalam dataset tanpa mengasumsikan kategori produk yang sebenarnya.

In [23]:
# Check Missing Values in Product by Dimension
dimension_columns = [
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"
]

products_clean[
    products_clean[dimension_columns].isnull().any(axis=1)
]

,product_id,product_category_name,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1803,zX9HL81jfvr2,toys,NaN,NaN,NaN,NaN
8274,VqyFCmRtRdfR,toys,NaN,NaN,NaN,NaN


Terdapat 2 produk yang tidak memiliki informasi berat dan dimensi. Karena produk tersebut masih memiliki informasi valid yang dapat digunakan untuk analisis transaksi, kategori, dan penjualan, record tetap dipertahankan dalam dataset utama.

Produk dengan missing values hanya akan dikecualikan pada analisis yang secara khusus membutuhkan informasi berat atau dimensi produk.

In [24]:
# Validation Numeric Values
numeric_checks = {
    "Order Items": (
        order_items_clean,
        ["price", "shipping_charges"]
    ),
    "Products": (
        products_clean,
        [
            "product_weight_g",
            "product_length_cm",
            "product_height_cm",
            "product_width_cm"
        ]
    ),
    "Payments": (
        payments_clean,
        ["payment_installments", "payment_value"]
    )
}

for name, (df, columns) in numeric_checks.items():
    print(f"\n{name}")

    for column in columns:
        print(
            f"{column}: "
            f"min={df[column].min()}, "
            f"max={df[column].max()}, "
            f"zero={(df[column] == 0).sum()}, "
            f"negative={(df[column] < 0).sum()}"
        )


Order Items
price: min=0.85, max=6735.0, zero=0, negative=0
shipping_charges: min=0.0, max=409.68, zero=22, negative=0

Products
product_weight_g: min=0.0, max=40425.0, zero=4, negative=0
product_length_cm: min=7.0, max=105.0, zero=0, negative=0
product_height_cm: min=2.0, max=105.0, zero=0, negative=0
product_width_cm: min=6.0, max=118.0, zero=0, negative=0

Payments
payment_installments: min=0, max=24, zero=3, negative=0
payment_value: min=0.0, max=7274.88, zero=1, negative=0


Tidak ditemukan nilai negatif pada `price` maupun `shipping_charges`. Terdapat 22 transaksi dengan nilai `shipping_charges` sebesar 0. Nilai tersebut tetap dipertahankan karena dapat merepresentasikan transaksi dengan bebas biaya pengiriman atau subsidi pengiriman.

In [25]:
# Check Numeric Values in Product by product_weight_g = 0
products_clean[
    products_clean["product_weight_g"] == 0
]

,product_id,product_category_name,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1298,fsq5lo34nHGK,toys,0.0,30.0,25.0,30.0
19935,aMbBkxpim3xt,toys,0.0,30.0,25.0,30.0
23122,nm5AMMVOx4Ac,toys,0.0,30.0,25.0,30.0
26485,xUlGNJzxYn1N,toys,0.0,30.0,25.0,30.0


Ditemukan 4 produk dengan berat sebesar 0 gram meskipun informasi dimensinya tersedia. Karena nilai tersebut tidak merepresentasikan berat produk yang realistis, nilai 0 dikategorikan sebagai missing value.

In [26]:
# Filtering product_weight_g in The "toys" Category to Exclude 0 Values
toys_weight_median = products_clean.loc[
    (products_clean["product_category_name"] == "toys") &
    (products_clean["product_weight_g"] > 0),
    "product_weight_g"
].median()

print("Median product weight category toys:", toys_weight_median)

Median product weight category toys: 700.0


In [27]:
# Imputing 0 Values in product_weight_g with The "toys" Category Median
products_clean.loc[
    (products_clean["product_category_name"] == "toys") &
    (products_clean["product_weight_g"] == 0),
    "product_weight_g"
] = toys_weight_median

In [28]:
# Check Results
print(
    "Jumlah product_weight_g = 0 pada kategori toys:",
    (
        (products_clean["product_category_name"] == "toys") &
        (products_clean["product_weight_g"] == 0)
    ).sum()
)

Jumlah product_weight_g = 0 pada kategori toys: 0


Ditemukan 4 produk pada kategori toys dengan berat sebesar 0 gram meskipun informasi dimensinya tersedia. Karena nilai tersebut tidak merepresentasikan berat produk yang realistis, nilai 0 diganti menggunakan median product_weight_g dari produk dalam kategori toys. Pendekatan ini digunakan agar nilai pengganti lebih merepresentasikan karakteristik produk sejenis tanpa menghapus record produk dari dataset.

In [29]:
# Check Numeric Values in Payments by payment_installments = 0
payments_clean[
    payments_clean["payment_installments"] == 0
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
74862,Xx1Ul7tUf1tJ,2,credit_card,0,1044.03
84085,piTvxOtCfbaE,2,credit_card,0,378.99
89054,hN1V5fL1VbER,2,credit_card,0,378.99


Ditemukan 3 transaksi credit card dengan `payment_installments` sebesar 0. Karena tidak tersedia informasi yang cukup untuk menentukan apakah nilai tersebut merupakan kesalahan input atau representasi pembayaran tanpa cicilan, nilai tetap dipertahankan dan akan diperhatikan secara khusus pada analisis installment.

In [30]:
# Check Numeric Values in Payments by payment_value = 0
payments_clean[
    payments_clean["payment_value"] == 0
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
65644,gZkNlHVVt1Xp,1,credit_card,8,0.0


Ditemukan 1 transaksi dengan `payment_value` sebesar 0. Record tetap dipertahankan karena informasi transaksi lainnya masih valid. Nilai tersebut akan diperlakukan sebagai anomali dan tidak digunakan sebagai dasar utama perhitungan nilai penjualan.

Hasil dari validasi numerik menunjukkan bahwa tidak terdapat nilai negatif pada variabel numerik utama. Beberapa nilai nol ditemukan dan dievaluasi berdasarkan konteks bisnisnya.

Nilai `shipping_charges` sebesar 0 tetap dipertahankan karena memungkinkan terjadinya free shipping. Sementara itu, nilai `product_weight_g` sebesar 0 dikategorikan sebagai missing value karena tidak realistis untuk produk fisik yang memiliki dimensi.

Nilai 0 pada `payment_installments` dan `payment_value` tetap dipertahankan karena tidak terdapat informasi yang cukup untuk menentukan nilai pengganti secara akurat.

In [31]:
# Check Relationship Between Dataset
relationship_checks = {
    "Orders customer_id tidak ditemukan di Customers":
        (~orders_clean["customer_id"].isin(customers_clean["customer_id"])).sum(),

    "Order Items order_id tidak ditemukan di Orders":
        (~order_items_clean["order_id"].isin(orders_clean["order_id"])).sum(),

    "Order Items product_id tidak ditemukan di Products":
        (~order_items_clean["product_id"].isin(products_clean["product_id"])).sum(),

    "Payments order_id tidak ditemukan di Orders":
        (~payments_clean["order_id"].isin(orders_clean["order_id"])).sum()
}

for check, result in relationship_checks.items():
    print(f"{check}: {result}")

Orders customer_id tidak ditemukan di Customers: 0
Order Items order_id tidak ditemukan di Orders: 0
Order Items product_id tidak ditemukan di Products: 0
Payments order_id tidak ditemukan di Orders: 0


Hasil validasi menunjukkan bahwa seluruh key antar dataset memiliki pasangan yang sesuai. Tidak ditemukan `customer_id`, `order_id`, maupun `product_id` yang kehilangan referensi pada tabel terkait.

Dengan demikian, hubungan antar dataset dapat digunakan dengan aman untuk proses penggabungan data dan analisis menggunakan SQL di Google BigQuery.

In [32]:
# Check Unique
print("Unique order_id pada Orders:",
      orders_clean["order_id"].nunique(),
      "dari", len(orders_clean))

print("Unique customer_id pada Customers:",
      customers_clean["customer_id"].nunique(),
      "dari", len(customers_clean))

print("Unique product_id pada Products:",
      products_clean["product_id"].nunique(),
      "dari", len(products_clean))

print("Unique order_id pada Payments:",
      payments_clean["order_id"].nunique(),
      "dari", len(payments_clean))

Unique order_id pada Orders: 89316 dari 89316
Unique customer_id pada Customers: 89316 dari 89316
Unique product_id pada Products: 27451 dari 27451
Unique order_id pada Payments: 89316 dari 89316


## Feature Engineering  

Pada tahap ini dibuat beberapa variabel baru berdasarkan data yang tersedia untuk mendukung analisis transaksi, waktu, dan performa pengiriman. Variabel baru dibuat tanpa mengubah informasi asli pada dataset.

In [33]:
# Create Time Feature in Orders
orders_clean["purchase_year"] = (
    orders_clean["order_purchase_timestamp"].dt.year
)

orders_clean["purchase_month"] = (
    orders_clean["order_purchase_timestamp"]
    .dt.to_period("M")
    .astype(str)
)

orders_clean["purchase_day"] = (
    orders_clean["order_purchase_timestamp"]
    .dt.day_name()
)

In [34]:
# Check Results
orders_clean[
    [
        "order_purchase_timestamp",
        "purchase_year",
        "purchase_month",
        "purchase_day"
    ]
].head()

,order_purchase_timestamp,purchase_year,purchase_month,purchase_day
0,2017-10-22 18:57:54,2017,2017-10,Sunday
1,2018-06-20 21:40:31,2018,2018-06,Wednesday
2,2018-02-16 16:19:31,2018,2018-02,Friday
3,2018-08-18 18:04:29,2018,2018-08,Saturday
4,2017-12-22 16:44:04,2017,2017-12,Friday


Variabel `purchase_year`, `purchase_month`, dan `purchase_day` dibuat dari `order_purchase_timestamp` untuk memudahkan analisis tren transaksi berdasarkan tahun, bulan, dan hari.

In [35]:
# Create Delivery Time
orders_clean["delivery_days"] = (
    orders_clean["order_delivered_timestamp"]
    - orders_clean["order_purchase_timestamp"]
).dt.total_seconds() / 86400

In [36]:
# Check Results
orders_clean[
    [
        "order_purchase_timestamp",
        "order_delivered_timestamp",
        "delivery_days"
    ]
].head()

,order_purchase_timestamp,order_delivered_timestamp,delivery_days
0,2017-10-22 18:57:54,2017-10-26 22:19:52,4.140255
1,2018-06-20 21:40:31,2018-07-03 22:51:22,13.049201
2,2018-02-16 16:19:31,2018-02-27 01:29:50,10.382164
3,2018-08-18 18:04:29,2018-08-27 20:03:51,9.082894
4,2017-12-22 16:44:04,2018-01-05 19:22:49,14.110243


Variabel `delivery_days` menunjukkan durasi aktual pengiriman sejak order dibuat hingga diterima oleh customer. Order yang belum memiliki waktu pengiriman tetap memiliki nilai kosong dan tidak digunakan dalam analisis durasi pengiriman.

In [37]:
# Create Delivery Status
orders_clean["delivery_status"] = np.select(
    [
        orders_clean["order_delivered_timestamp"].isna(),
        orders_clean["order_delivered_timestamp"]
        <= orders_clean["order_estimated_delivery_date"]
    ],
    [
        "Not Delivered",
        "On Time"
    ],
    default="Late"
)

In [38]:
# Check Results
orders_clean["delivery_status"].value_counts()

,count
delivery_status,
On Time,80689
Late,6738
Not Delivered,1889


Variabel `delivery_status` dibuat untuk mengelompokkan performa pengiriman menjadi `On Time`, `Late`, dan `Not Delivered`.

Pesanan dikategorikan sebagai `On Time` apabila waktu pengiriman aktual tidak melewati estimasi pengiriman, `Late` apabila melewati estimasi, dan `Not Delivered` apabila belum memiliki waktu pengiriman aktual.

In [39]:
# Create Order Values
order_items_clean["order_value"] = (
    order_items_clean["price"]
    + order_items_clean["shipping_charges"]
)

In [40]:
# Check Results
order_items_clean[
    [
        "price",
        "shipping_charges",
        "order_value"
    ]
].head()

,price,shipping_charges,order_value
0,223.51,84.65,308.16
1,170.80,23.79,194.59
2,64.40,17.38,81.78
3,264.50,30.72,295.22
4,779.90,30.66,810.56


Variabel `order_value` dihitung dari penjumlahan harga produk (`price`) dan biaya pengiriman (`shipping_charges`). Variabel ini akan digunakan untuk melihat hasil transaksi setiap order.

Sementara itu, `price` tetap digunakan sebagai nilai penjualan produk agar biaya pengiriman dapat dianalisis secara terpisah.

## 8. Final Data Validation

Pada tahap ini dilakukan pemeriksaan akhir terhadap dataset setelah proses data cleaning dan feature engineering. Validasi dilakukan untuk memastikan tidak terdapat duplikasi yang tidak diharapkan, hubungan antar dataset tetap konsisten, serta variabel baru memiliki nilai yang sesuai sebelum dataset digunakan pada tahap analisis SQL.

In [41]:
# Check Final Data Validation
clean_datasets = {
    "Orders": orders_clean,
    "Order Items": order_items_clean,
    "Products": products_clean,
    "Customers": customers_clean,
    "Payments": payments_clean
}

for name, df in clean_datasets.items():
    print(f"{name}: {df.shape[0]} baris dan {df.shape[1]} kolom")

Orders: 89316 baris dan 12 kolom
Order Items: 89316 baris dan 6 kolom
Products: 27451 baris dan 6 kolom
Customers: 89316 baris dan 4 kolom
Payments: 89316 baris dan 5 kolom


In [42]:
# Check Final Data Duplicate
for name, df in clean_datasets.items():
    print(f"{name}: {df.duplicated().sum()} duplicate rows")

Orders: 0 duplicate rows
Order Items: 0 duplicate rows
Products: 0 duplicate rows
Customers: 0 duplicate rows
Payments: 0 duplicate rows


In [43]:
for name, df in clean_datasets.items():
    print(f"\n{name}")

    missing = df.isnull().sum()
    print(missing[missing > 0])
    print(f"{'='*50}")


Orders
order_approved_at               9
order_delivered_timestamp    1889
delivery_days                1889
dtype: int64

Order Items
Series([], dtype: int64)

Products
product_weight_g     2
product_length_cm    2
product_height_cm    2
product_width_cm     2
dtype: int64

Customers
Series([], dtype: int64)

Payments
Series([], dtype: int64)


In [44]:
# Export File
orders_clean.to_csv("orders_clean.csv", index=False)
order_items_clean.to_csv("order_items_clean.csv", index=False)
products_clean.to_csv("products_clean.csv", index=False)
customers_clean.to_csv("customers_clean.csv", index=False)
payments_clean.to_csv("payments_clean.csv", index=False)

In [45]:
# Import Library
from google.colab import auth, data_table
from google.cloud import bigquery
from pandas_gbq import to_gbq

# Proces Aunthentication
auth.authenticate_user()
print('Authenticated')

Authenticated


In [46]:
# Access Dataset in BigQuery
project_id = 'projek-muhroni-dqlab'
client = bigquery.Client(project = project_id)
print('Access Denied')

Access Denied


## SQL Data Validation

In [47]:
%%bigquery --project {project_id} --verbose

# Check Dataset in BigQuery
SELECT * FROM ecommerce.orders LIMIT 5

Executing query with job ID: 377968f2-65c7-4581-b961-871cff90e092
Query executing: 0.67s
Job ID 377968f2-65c7-4581-b961-871cff90e092 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_timestamp,order_estimated_delivery_date,purchase_year,purchase_month,purchase_day,delivery_days,delivery_status
0,rmbo1tI4QSfi,FM5xLRCCs67e,shipped,2016-09-04 21:15:19+00:00,2016-10-07 13:18:03+00:00,NaT,2016-10-20,2016,2016-09,Sunday,NaN,Not Delivered
1,HfwGXKFBqK8v,9eIHDJ0EfBR3,shipped,2016-09-04 21:15:19+00:00,2016-10-07 13:18:03+00:00,NaT,2016-10-20,2016,2016-09,Sunday,NaN,Not Delivered
2,2oxjQVD8TlnM,jCHyXX2AzdY4,canceled,2016-10-02 22:07:52+00:00,2016-10-06 15:50:56+00:00,NaT,2016-10-25,2016,2016-10,Sunday,NaN,Not Delivered
3,bwH0ju1ZxDr1,DYk8BhnWnrxA,delivered,2016-10-03 09:44:50+00:00,2016-10-06 15:50:54+00:00,2016-10-26 14:02:13+00:00,2016-10-27,2016,2016-10,Monday,23.178738,On Time
4,g9U5B9yNEoB9,5Q073FPfjhDZ,delivered,2016-10-03 09:44:50+00:00,2016-10-06 15:50:54+00:00,2016-10-26 14:02:13+00:00,2016-10-27,2016,2016-10,Monday,23.178738,On Time


In [48]:
%%bigquery --project {project_id} --verbose

# Dataset Validation
SELECT
  'orders' AS table_name,
  COUNT(*) AS total_rows
FROM ecommerce.orders

UNION ALL

SELECT
  'order_items',
  COUNT(*)
FROM ecommerce.order_items

UNION ALL

SELECT
  'products',
  COUNT(*)
FROM ecommerce.products

UNION ALL

SELECT
  'customers',
  COUNT(*)
FROM ecommerce.customers

UNION ALL

SELECT
  'payments',
  COUNT(*)
FROM ecommerce.payments;

Executing query with job ID: 682d3d6e-10d8-4549-9920-265855e10952
Query executing: 0.64s
Job ID 682d3d6e-10d8-4549-9920-265855e10952 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,table_name,total_rows
0,orders,89316
1,order_items,89316
2,products,27451
3,customers,89316
4,payments,89316


In [49]:
%%bigquery --project {project_id} --verbose

# Check Primary Key in Table Orders
SELECT
  COUNT(*) AS total_rows,
  COUNT(DISTINCT order_id) AS unique_order_id
FROM ecommerce.orders

Executing query with job ID: 4795185f-b813-4bf3-89e8-48caa29f6605
Query executing: 0.43s
Job ID 4795185f-b813-4bf3-89e8-48caa29f6605 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,total_rows,unique_order_id
0,89316,89316


In [50]:
%%bigquery --project {project_id} --verbose

# Check Primary Key in Table Products
SELECT
  COUNT(*) AS total_rows,
  COUNT(DISTINCT product_id) AS unique_product_id
FROM ecommerce.products

Executing query with job ID: 82e84234-1fc5-4810-b141-8302fe3e95ed
Query executing: 0.49s
Job ID 82e84234-1fc5-4810-b141-8302fe3e95ed successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,total_rows,unique_product_id
0,27451,27451


## Overall Business Performance  

How is the overall performance of the e-commerce business?

In [51]:
%%bigquery --project {project_id} --verbose

# Overall Business Performance by Status Category Delivered & Canceled
SELECT
  COUNT(DISTINCT o.order_id) AS total_orders,

  COUNTIF(o.order_status = 'delivered') AS delivered_orders,

  COUNTIF(o.order_status = 'canceled') AS canceled_orders,

  ROUND(
    SUM(
      CASE
        WHEN o.order_status = 'delivered'
        THEN oi.price
      END
    ), 2
  ) AS total_product_sales,

  ROUND(
    SUM(
      CASE
        WHEN o.order_status = 'delivered'
        THEN oi.shipping_charges
      END
    ), 2
  ) AS total_shipping_charges,

  ROUND(
    SUM(
      CASE
        WHEN o.order_status = 'delivered'
        THEN oi.order_value
      END
    ), 2
  ) AS total_order_value,

  ROUND(
    AVG(
      CASE
        WHEN o.order_status = 'delivered'
        THEN oi.order_value
      END
    ), 2
  ) AS average_order_value,

  ROUND(
    AVG(
      CASE
        WHEN o.order_status = 'delivered'
        THEN o.delivery_days
      END
    ), 2
  ) AS average_delivery_days

FROM ecommerce.orders AS o

LEFT JOIN ecommerce.order_items AS oi
  ON o.order_id = oi.order_id;

Executing query with job ID: 7e0ef57d-74d0-43ef-87a3-12db2a02b53c
Query executing: 0.85s
Job ID 7e0ef57d-74d0-43ef-87a3-12db2a02b53c successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,total_orders,delivered_orders,canceled_orders,total_product_sales,total_shipping_charges,total_order_value,average_order_value,average_delivery_days
0,89316,87428,409,29925513.630000000,3868389.390000000,33793903.020000000,386.530000000,12.43


## Analysis Product Sales & Order Trends  

1. How have revenue and the number of orders changed each month?  
2. Have sales increased or decreased compared to the previous month?  
3. Which day are customers most active in making transactions?
4. How did the e-commerce business performance change year over year?
5. Which product categories generate the highest sales and order volume?
6. How much does each product category contribute to total product sales?

In [52]:
%%bigquery --project {project_id} --verbose

# Monthly Sales Performance
WITH monthly_order_data AS (
  SELECT
    o.order_id,
    FORMAT_DATE('%Y-%m', o.order_purchase_timestamp) AS month_key,
    oi.price,
    oi.order_value,
    o.order_status
  FROM ecommerce.orders AS o
  LEFT JOIN ecommerce.order_items AS oi
    ON o.order_id = oi.order_id
)
SELECT
  FORMAT_DATE('%Y - %B', PARSE_DATE('%Y-%m', month_key)) AS purchase_month,
  COUNT(DISTINCT order_id) AS total_orders,
  ROUND(
    SUM(price),
    2
  ) AS total_sales,
  ROUND(
    AVG(order_value),
    2
  ) AS average_order_value
FROM monthly_order_data
WHERE order_status = 'delivered'
GROUP BY
  month_key
ORDER BY
  PARSE_DATE('%Y-%m', month_key);

Executing query with job ID: 7adf5379-23f1-4583-96d8-1369acff4fb8
Query executing: 0.61s
Job ID 7adf5379-23f1-4583-96d8-1369acff4fb8 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,purchase_month,total_orders,total_sales,average_order_value
0,2016 - October,256,87453.550000000,389.450000000
1,2016 - December,1,215.940000000,277.260000000
2,2017 - January,824,231539.380000000,327.440000000
3,2017 - February,1467,426872.750000000,337.220000000
4,2017 - March,2316,799208.640000000,393.200000000
5,2017 - April,2042,701127.960000000,389.040000000
6,2017 - May,3301,1007240.810000000,351.450000000
7,2017 - June,2884,896081.470000000,354.320000000
8,2017 - July,3699,1190242.860000000,366.780000000
9,2017 - August,3942,1251579.030000000,362.060000000


In [53]:
%%bigquery --project {project_id} --verbose

# Monthly Sales Growth
WITH monthly_sales AS (

SELECT
  FORMAT_DATE('%Y-%m', o.order_purchase_timestamp) AS month_key,

  SUM(oi.price) AS total_sales

FROM ecommerce.orders AS o

LEFT JOIN ecommerce.order_items AS oi
ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY
  month_key

)

SELECT

  FORMAT_DATE('%Y - %B', PARSE_DATE('%Y-%m', month_key)) AS purchase_month,

  ROUND(total_sales,2) AS total_sales,

  ROUND(
    LAG(total_sales)
    OVER(
      ORDER BY PARSE_DATE('%Y-%m', month_key)
    ),
    2
  ) AS previous_month_sales,


  ROUND(
    SAFE_DIVIDE(
      total_sales -
      LAG(total_sales)
      OVER(
        ORDER BY PARSE_DATE('%Y-%m', month_key)
      ),

      LAG(total_sales)
      OVER(
        ORDER BY PARSE_DATE('%Y-%m', month_key)
      )

    ) * 100,
    2
  ) AS growth_percentage


FROM monthly_sales

ORDER BY PARSE_DATE('%Y-%m', month_key);

Executing query with job ID: 29dab15f-3c55-49cd-ae18-8482a498f197
Query executing: 0.55s
Job ID 29dab15f-3c55-49cd-ae18-8482a498f197 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,purchase_month,total_sales,previous_month_sales,growth_percentage
0,2016 - October,87453.550000000,None,None
1,2016 - December,215.940000000,87453.550000000,-99.750000000
2,2017 - January,231539.380000000,215.940000000,107123.940000000
3,2017 - February,426872.750000000,231539.380000000,84.360000000
4,2017 - March,799208.640000000,426872.750000000,87.220000000
5,2017 - April,701127.960000000,799208.640000000,-12.270000000
6,2017 - May,1007240.810000000,701127.960000000,43.660000000
7,2017 - June,896081.470000000,1007240.810000000,-11.040000000
8,2017 - July,1190242.860000000,896081.470000000,32.830000000
9,2017 - August,1251579.030000000,1190242.860000000,5.150000000


Monthly Sales Performance

Analisis bulanan dilakukan untuk melihat perkembangan transaksi dan nilai penjualan dari waktu ke waktu.

Berdasarkan hasil analisis, performa penjualan menunjukkan pola perubahan setiap bulan yang dapat digunakan untuk memahami tren pertumbuhan, periode dengan transaksi tertinggi, serta potensi adanya periode dengan performa yang sedang menurun.

In [54]:
%%bigquery --project {project_id} --verbose

# Day of Week Performance
SELECT

  purchase_day,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales


FROM ecommerce.orders AS o

LEFT JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id


WHERE order_status = 'delivered'


GROUP BY
  purchase_day


ORDER BY
  total_orders DESC;

Executing query with job ID: d1521c9c-5e5c-4df8-81d2-4c9a47e457df
Query executing: 0.57s
Job ID d1521c9c-5e5c-4df8-81d2-4c9a47e457df successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,purchase_day,total_orders,total_sales
0,Tuesday,14301,4675114.620000000
1,Monday,14086,4702999.830000000
2,Wednesday,13631,4721401.360000000
3,Thursday,13160,4546653.930000000
4,Friday,12519,4489367.770000000
5,Sunday,10340,3560560.070000000
6,Saturday,9391,3229416.050000000


In [55]:
%%bigquery --project {project_id} --verbose

# Yearly Sales Performance
SELECT

  purchase_year,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales


FROM ecommerce.orders AS o

LEFT JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id


WHERE order_status = 'delivered'


GROUP BY
  purchase_year


ORDER BY
  purchase_year;

Executing query with job ID: b017393a-152f-451b-b33d-edcaa923bfa7
Query executing: 0.48s
Job ID b017393a-152f-451b-b33d-edcaa923bfa7 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,purchase_year,total_orders,total_sales
0,2016,257,87669.490000000
1,2017,40082,13897814.860000000
2,2018,47089,15940029.280000000


In [56]:
%%bigquery --project {project_id} --verbose

# Product Category Performance
SELECT
  p.product_category_name,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales,

  ROUND(
    AVG(oi.price),
    2
  ) AS average_product_price,

  ROUND(
    AVG(oi.order_value),
    2
  ) AS average_order_value

FROM ecommerce.orders AS o

LEFT JOIN ecommerce.order_items AS oi
  ON o.order_id = oi.order_id

LEFT JOIN ecommerce.products AS p
  ON oi.product_id = p.product_id

WHERE o.order_status = 'delivered'

GROUP BY
  p.product_category_name

ORDER BY
  total_sales DESC;


Executing query with job ID: b7582bd7-8454-48d9-8219-e7a5b3d85f0c
Query executing: 0.65s
Job ID b7582bd7-8454-48d9-8219-e7a5b3d85f0c successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,product_category_name,total_orders,total_sales,average_product_price,average_order_value
0,toys,65643,22613413.450000000,344.490000000,389.020000000
1,furniture_decor,1723,739082.570000000,428.950000000,473.470000000
2,garden_tools,777,732447.160000000,942.660000000,984.160000000
3,bed_bath_table,2119,594526.860000000,280.570000000,324.730000000
4,health_beauty,2289,581728.490000000,254.140000000,297.090000000
...,...,...,...,...,...
66,flowers,6,231.720000000,38.620000000,76.520000000
67,furniture_mattress_and_upholstery,1,76.990000000,76.990000000,93.980000000
68,fashio_female_clothing,2,68.140000000,34.070000000,78.480000000
69,security_and_services,1,45.900000000,45.900000000,52.210000000


In [57]:
%%bigquery --project {project_id} --verbose

# Top 10 Product Categories by Sales

SELECT
  p.product_category_name,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi
  ON o.order_id = oi.order_id

JOIN ecommerce.products AS p
  ON oi.product_id = p.product_id

WHERE o.order_status = 'delivered'

GROUP BY
  p.product_category_name

ORDER BY
  total_sales DESC

LIMIT 10;

Executing query with job ID: 835ec945-e6e7-408f-86f2-c98f1f18d1b6
Query executing: 0.64s
Job ID 835ec945-e6e7-408f-86f2-c98f1f18d1b6 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,product_category_name,total_orders,total_sales
0,toys,65643,22613413.450000000
1,furniture_decor,1723,739082.570000000
2,garden_tools,777,732447.160000000
3,bed_bath_table,2119,594526.860000000
4,health_beauty,2289,581728.490000000
5,watches_gifts,1177,577406.420000000
6,sports_leisure,1799,530014.030000000
7,telephony,891,491089.470000000
8,computers_accessories,1673,423136.320000000
9,housewares,1301,406031.190000000


In [58]:
%%bigquery --project {project_id} --verbose

# Sales Contribution by Category

WITH category_sales AS (

  SELECT
    p.product_category_name,
    COUNT(DISTINCT o.order_id) AS total_orders,
    SUM(oi.price) AS total_sales

  FROM ecommerce.orders AS o

  JOIN ecommerce.order_items AS oi
    ON o.order_id = oi.order_id

  JOIN ecommerce.products AS p
    ON oi.product_id = p.product_id

  WHERE o.order_status = 'delivered'

  GROUP BY
    p.product_category_name
)

SELECT
  product_category_name,
  total_orders,

  ROUND(
    total_sales,
    2
  ) AS total_sales,

  ROUND(
    SAFE_DIVIDE(
      total_sales,
      SUM(total_sales) OVER()
    ) * 100,
    2
  ) AS sales_contribution_percent

FROM category_sales

ORDER BY
  total_sales DESC;

Executing query with job ID: 75f14014-e970-473a-ae8a-d1c86c36c1d6
Query executing: 0.68s
Job ID 75f14014-e970-473a-ae8a-d1c86c36c1d6 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,product_category_name,total_orders,total_sales,sales_contribution_percent
0,toys,65643,22613413.450000000,75.570000000
1,furniture_decor,1723,739082.570000000,2.470000000
2,garden_tools,777,732447.160000000,2.450000000
3,bed_bath_table,2119,594526.860000000,1.990000000
4,health_beauty,2289,581728.490000000,1.940000000
...,...,...,...,...
66,flowers,6,231.720000000,0E-9
67,furniture_mattress_and_upholstery,1,76.990000000,0E-9
68,fashio_female_clothing,2,68.140000000,0E-9
69,security_and_services,1,45.900000000,0E-9


In [59]:
%%bigquery --project {project_id} --verbose

# Category Ranking
WITH category_performance AS (

  SELECT
    p.product_category_name,

    COUNT(DISTINCT o.order_id) AS total_orders,

    SUM(oi.price) AS total_sales

  FROM `projek-muhroni-dqlab.ecommerce.orders` o

  JOIN `projek-muhroni-dqlab.ecommerce.order_items` oi
    ON o.order_id = oi.order_id

  JOIN `projek-muhroni-dqlab.ecommerce.products` p
    ON oi.product_id = p.product_id

  WHERE o.order_status = 'delivered'

  GROUP BY
    p.product_category_name
)

SELECT
  product_category_name,

  total_orders,

  ROUND(total_sales, 2) AS total_sales,

  RANK() OVER(
    ORDER BY total_sales DESC
  ) AS sales_rank,

  RANK() OVER(
    ORDER BY total_orders DESC
  ) AS order_volume_rank

FROM category_performance

ORDER BY
  sales_rank;

Executing query with job ID: 697713a5-ac4f-4700-a3d3-a1dd9f9e40f5
Query executing: 0.52s
Job ID 697713a5-ac4f-4700-a3d3-a1dd9f9e40f5 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,product_category_name,total_orders,total_sales,sales_rank,order_volume_rank
0,toys,65643,22613413.450000000,1,1
1,furniture_decor,1723,739082.570000000,2,5
2,garden_tools,777,732447.160000000,3,11
3,bed_bath_table,2119,594526.860000000,4,3
4,health_beauty,2289,581728.490000000,5,2
...,...,...,...,...,...
66,flowers,6,231.720000000,67,61
67,furniture_mattress_and_upholstery,1,76.990000000,68,68
68,fashio_female_clothing,2,68.140000000,69,66
69,security_and_services,1,45.900000000,70,68


## Delivery & Logistics Performance Analysis  

1. How well is the delivery service performing?  
2. What percentage of completed deliveries were delayed?  
3. How much longer does delivery take compared to customer expectation?
4. Which regions experience the highest delivery problems?  
5. Is delivery performance improving or getting worse over time?

In [60]:
%%bigquery --project {project_id} --verbose

# Overall Delivery Performance
SELECT

  delivery_status,

  COUNT(order_id) AS total_orders,

  ROUND(
    AVG(delivery_days),
    2
  ) AS average_delivery_days

FROM `projek-muhroni-dqlab.ecommerce.orders`

GROUP BY
  delivery_status

ORDER BY
  total_orders DESC;

Executing query with job ID: f63ad126-fdae-4dbb-8a76-55d52a8ed884
Query executing: 0.42s
Job ID f63ad126-fdae-4dbb-8a76-55d52a8ed884 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,delivery_status,total_orders,average_delivery_days
0,On Time,80689,10.86
1,Late,6738,31.32
2,Not Delivered,1889,NaN


Delivery Performance Analysis

Analisis delivery dilakukan untuk mengevaluasi efektivitas proses pengiriman berdasarkan status pengiriman, durasi pengiriman aktual, dan tingkat keterlambatan dibandingkan estimasi waktu pengiriman.

Analisis menggunakan variabel `delivery_days` dan `delivery_status` yang telah dibuat pada tahap feature engineering.

In [61]:
%%bigquery --project {project_id} --verbose

# Late Delivery Rate

SELECT

  COUNTIF(delivery_status = 'Late')
  AS late_orders,

  COUNTIF(
    delivery_status IN ('Late','On Time')
  )
  AS completed_deliveries,

  ROUND(

    SAFE_DIVIDE(
      COUNTIF(delivery_status = 'Late'),

      COUNTIF(
        delivery_status IN ('Late','On Time')
      )

    ) * 100,

    2

  ) AS late_delivery_rate

FROM `projek-muhroni-dqlab.ecommerce.orders`;

Executing query with job ID: d3278d82-3e3d-4d09-a486-53102962e77a
Query executing: 0.43s
Job ID d3278d82-3e3d-4d09-a486-53102962e77a successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,late_orders,completed_deliveries,late_delivery_rate
0,6738,87427,7.71


In [62]:
%%bigquery --project {project_id} --verbose

# Actual Delivery vs Estimated Delivery
SELECT

  ROUND(
    AVG(
      delivery_days
    ),
    2
  ) AS actual_delivery_days,


  ROUND(
    AVG(
      DATE_DIFF(
        order_estimated_delivery_date,
        DATE(order_purchase_timestamp),
        DAY
      )
    ),
    2
  ) AS estimated_delivery_days

FROM ecommerce.orders

WHERE
  order_status = 'delivered';

Executing query with job ID: b58ddbfd-411a-40ed-aa51-5a23e3cf026b
Query executing: 0.45s
Job ID b58ddbfd-411a-40ed-aa51-5a23e3cf026b successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,actual_delivery_days,estimated_delivery_days
0,12.43,24.5


In [63]:
%%bigquery --project {project_id} --verbose

# Delivery Performance by State
SELECT

  c.customer_state,

  COUNT(o.order_id) AS total_orders,

  COUNTIF(
    o.delivery_status = 'Late'
  ) AS late_orders,


  ROUND(

    SAFE_DIVIDE(
      COUNTIF(
        o.delivery_status = 'Late'
      ),
      COUNT(o.order_id)

    ) * 100,

    2

  ) AS late_delivery_rate,


  ROUND(
    AVG(o.delivery_days),
    2
  ) AS avg_delivery_days


FROM ecommerce.orders AS o


JOIN ecommerce.customers AS c

ON o.customer_id = c.customer_id


WHERE
  o.delivery_status != 'Not Delivered'


GROUP BY
  c.customer_state


ORDER BY
  late_delivery_rate DESC;

Executing query with job ID: 80ba3b85-930b-4abd-bcaf-bf40ad2a21b1
Query executing: 0.45s
Job ID 80ba3b85-930b-4abd-bcaf-bf40ad2a21b1 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,customer_state,total_orders,late_orders,late_delivery_rate,avg_delivery_days
0,AL,325,79,24.31,24.71
1,MA,603,125,20.73,21.29
2,PI,409,67,16.38,19.02
3,SE,297,47,15.82,21.21
4,CE,1101,167,15.17,20.70
5,BA,3014,421,13.97,19.33
6,TO,243,31,12.76,17.34
7,RJ,11251,1428,12.69,15.21
8,PA,799,93,11.64,23.24
9,ES,1777,206,11.59,15.57


In [64]:
%%bigquery --project {project_id} --verbose

# Delivery Trend Over Time
WITH delivery_trend AS (

SELECT

  DATE_TRUNC(
    DATE(order_purchase_timestamp),
    MONTH
  ) AS month_date,

  FORMAT_DATE(
    '%Y-%B',
    DATE(order_purchase_timestamp)
  ) AS purchase_month,

  COUNT(order_id) AS total_orders,

  ROUND(
    SAFE_DIVIDE(
      COUNTIF(delivery_status = 'Late'),
      COUNTIF(delivery_status IN ('Late','On Time'))
    ) * 100,
    2
  ) AS late_delivery_rate


FROM ecommerce.orders`

GROUP BY
  month_date,
  purchase_month

)

SELECT

  purchase_month,
  total_orders,
  late_delivery_rate

FROM delivery_trend

ORDER BY
  month_date ASC;

Executing query with job ID: 8ee6d128-aafb-4d5c-b49c-9e8d257b08eb
Query executing: 0.11s


ERROR:
 400 Syntax error: Unclosed identifier literal at [27:22]; reason: invalidQuery, location: query, message: Syntax error: Unclosed identifier literal at [27:22]

Location: US
Job ID: 8ee6d128-aafb-4d5c-b49c-9e8d257b08eb



## Geographic Performance Analysis  

1. Which states have the highest number of customers and orders?
2. Which states generate the highest sales contribution?
3. How much does each state contribute to total sales?
4. Which regions have customers with higher spending behavior?
5. Which regions experience delivery challenges?

In [65]:
%%bigquery --project {project_id} --verbose

# Customer Distribution by State
SELECT

  c.customer_state,

  COUNT(DISTINCT c.customer_id) AS total_customers,

  COUNT(DISTINCT o.order_id) AS total_orders


FROM ecommerce.customers AS c

JOIN ecommerce.orders AS o

ON c.customer_id = o.customer_id


WHERE
  o.order_status = 'delivered'

GROUP BY
  c.customer_state

ORDER BY
  total_orders DESC;

Executing query with job ID: 898004cc-4099-464d-b49d-8f9d699a065b
Query executing: 0.53s
Job ID 898004cc-4099-464d-b49d-8f9d699a065b successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,customer_state,total_customers,total_orders
0,SP,37101,37101
1,RJ,11249,11249
2,MG,10172,10172
3,RS,4854,4854
4,PR,4451,4451
5,SC,3160,3160
6,BA,3014,3014
7,GO,1816,1816
8,ES,1777,1777
9,DF,1755,1755


In [66]:
%%bigquery --project {project_id} --verbose

# Sales Performance by State

SELECT

  c.customer_state,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales,

  ROUND(
    AVG(oi.order_value),
    2
  ) AS average_order_value

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

JOIN ecommerce.customers AS c

ON o.customer_id = c.customer_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  c.customer_state

ORDER BY
  total_sales DESC;

Executing query with job ID: ba54fca7-980c-4984-ac36-236c483d87c9
Query executing: 0.46s
Job ID ba54fca7-980c-4984-ac36-236c483d87c9 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,customer_state,total_orders,total_sales,average_order_value
0,SP,37101,12740951.710000000,386.920000000
1,RJ,11249,3755780.990000000,376.470000000
2,MG,10172,3465654.460000000,386.250000000
3,RS,4854,1645269.930000000,384.180000000
4,PR,4451,1580718.970000000,399.590000000
5,SC,3160,1051289.750000000,377.120000000
6,BA,3014,1041864.980000000,392.310000000
7,GO,1816,666349.990000000,411.640000000
8,ES,1777,611893.080000000,387.990000000
9,DF,1755,559149.880000000,360.170000000


In [67]:
%%bigquery --project {project_id} --verbose

# Sales Contribution by State

WITH state_sales AS (

SELECT

  c.customer_state,

  SUM(oi.price) AS total_sales

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

JOIN ecommerce.customers AS c

ON o.customer_id = c.customer_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  c.customer_state

)

SELECT

  customer_state,

  ROUND(total_sales,2) AS total_sales,

  ROUND(

    SAFE_DIVIDE(
      total_sales,
      SUM(total_sales) OVER()
    ) * 100,

    2

  ) AS sales_contribution_percent

FROM state_sales

ORDER BY
  total_sales DESC;

Executing query with job ID: 91d9d7c7-b9d0-4b34-a436-7504d858c69e
Query executing: 0.46s
Job ID 91d9d7c7-b9d0-4b34-a436-7504d858c69e successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,customer_state,total_sales,sales_contribution_percent
0,SP,12740951.710000000,42.580000000
1,RJ,3755780.990000000,12.550000000
2,MG,3465654.460000000,11.580000000
3,RS,1645269.930000000,5.500000000
4,PR,1580718.970000000,5.280000000
5,SC,1051289.750000000,3.510000000
6,BA,1041864.980000000,3.480000000
7,GO,666349.990000000,2.230000000
8,ES,611893.080000000,2.040000000
9,DF,559149.880000000,1.870000000


In [68]:
%%bigquery --project {project_id} --verbose

# Average Order Value by Region

SELECT

  c.customer_state,

  ROUND(
    AVG(oi.order_value),
    2
  ) AS average_order_value,

  COUNT(DISTINCT o.order_id) AS total_orders

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

JOIN ecommerce.customers AS c

ON o.customer_id = c.customer_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  c.customer_state

HAVING
  total_orders >= 100

ORDER BY
  average_order_value DESC;

Executing query with job ID: 9f4b78c2-6808-48fe-adcc-1a756e2bef27
Query executing: 0.43s
Job ID 9f4b78c2-6808-48fe-adcc-1a756e2bef27 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,customer_state,average_order_value,total_orders
0,RO,479.750000000,237
1,PA,421.440000000,799
2,GO,411.640000000,1816
3,CE,411.080000000,1101
4,PB,409.970000000,480
5,PE,409.090000000,1445
6,RN,408.090000000,377
7,PR,399.590000000,4451
8,MA,394.990000000,603
9,TO,392.490000000,243


In [69]:
%%bigquery --project {project_id} --verbose

# Regional Delivery Performance

SELECT

  c.customer_state,

  COUNT(o.order_id) AS total_orders,

  COUNTIF(
    o.delivery_status = 'Late'
  ) AS late_orders,

  ROUND(

    SAFE_DIVIDE(
      COUNTIF(
        o.delivery_status = 'Late'
      ),
      COUNTIF(
        o.delivery_status IN ('Late','On Time')
      )
    ) * 100,

    2

  ) AS late_delivery_rate

FROM ecommerce.orders AS o

JOIN ecommerce.customers AS c

ON o.customer_id = c.customer_id

WHERE
  o.delivery_status != 'Not Delivered'

GROUP BY
  c.customer_state

ORDER BY
  late_delivery_rate DESC;

Executing query with job ID: 20e57042-b44f-4d61-9acd-0f6012eb291b
Query executing: 0.34s
Job ID 20e57042-b44f-4d61-9acd-0f6012eb291b successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,customer_state,total_orders,late_orders,late_delivery_rate
0,AL,325,79,24.31
1,MA,603,125,20.73
2,PI,409,67,16.38
3,SE,297,47,15.82
4,CE,1101,167,15.17
5,BA,3014,421,13.97
6,TO,243,31,12.76
7,RJ,11251,1428,12.69
8,PA,799,93,11.64
9,ES,1777,206,11.59


## Seller Performance Analysis  

1. Which sellers generate the highest sales revenue?
2. Who are the top revenue-generating sellers?
3. How much does each seller contribute to total sales?
4. Which sellers handle the highest transaction volume?
5. Which sellers have higher delivery issues?

In [70]:
%%bigquery --project {project_id} --verbose

# Seller Sales Performance Analysis

SELECT

  oi.seller_id,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales,

  ROUND(
    AVG(oi.order_value),
    2
  ) AS average_order_value

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  oi.seller_id

ORDER BY
  total_sales DESC;

Executing query with job ID: 0341931d-79cd-437f-901e-10878f18507d
Query executing: 0.55s
Job ID 0341931d-79cd-437f-901e-10878f18507d successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,seller_id,total_orders,total_sales,average_order_value
0,r7VxefoTVHbb,1512,2562407.580000000,1735.360000000
1,yGbPyLPc8PmT,1586,1900070.900000000,1237.930000000
2,K0qPVGdA91KO,1151,663779.980000000,622.110000000
3,l1pYW6GBnPMr,905,474665.710000000,560.660000000
4,ruq0u2ZpAMDr,1284,419203.030000000,365.890000000
...,...,...,...,...
2817,iAF0NxaypyZB,1,6.190000000,15.820000000
2818,dzTjh3Rd62QN,1,0.850000000,10.340000000
2819,WHCTdImrIp0y,1,0.850000000,3.310000000
2820,uzHNB354dHwV,1,0.850000000,71.670000000


In [71]:
%%bigquery --project {project_id} --verbose

# Top 10 Sellers by Revenue

SELECT

  oi.seller_id,

  COUNT(DISTINCT o.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  oi.seller_id

ORDER BY
  total_sales DESC

LIMIT 10;

Executing query with job ID: 36890e84-2ab6-4f41-a361-bab5865d2727
Query executing: 0.57s
Job ID 36890e84-2ab6-4f41-a361-bab5865d2727 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,seller_id,total_orders,total_sales
0,r7VxefoTVHbb,1512,2562407.580000000
1,yGbPyLPc8PmT,1586,1900070.900000000
2,K0qPVGdA91KO,1151,663779.980000000
3,l1pYW6GBnPMr,905,474665.710000000
4,ruq0u2ZpAMDr,1284,419203.030000000
5,RKad98cTxhSb,1624,361091.930000000
6,EGW4UK5bOeEZ,1066,304213.080000000
7,yHsOFiqVrV1u,866,292845.580000000
8,BWczKEOZMEFy,300,278055.100000000
9,coi9xkdo66JI,869,238745.040000000


In [72]:
%%bigquery --project {project_id} --verbose

# Seller Contribution Percentage Analysis

WITH seller_sales AS (

SELECT

  oi.seller_id,

  SUM(oi.price) AS total_sales

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  oi.seller_id

)

SELECT

  seller_id,

  ROUND(
    total_sales,
    2
  ) AS total_sales,

  ROUND(

    SAFE_DIVIDE(
      total_sales,
      SUM(total_sales) OVER()
    ) * 100,

    2

  ) AS sales_contribution_percent

FROM seller_sales

ORDER BY
  total_sales DESC;

Executing query with job ID: 22da8f81-c6d0-493c-9f96-28c0cd2e5fc6
Query executing: 0.37s
Job ID 22da8f81-c6d0-493c-9f96-28c0cd2e5fc6 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,seller_id,total_sales,sales_contribution_percent
0,r7VxefoTVHbb,2562407.580000000,8.560000000
1,yGbPyLPc8PmT,1900070.900000000,6.350000000
2,K0qPVGdA91KO,663779.980000000,2.220000000
3,l1pYW6GBnPMr,474665.710000000,1.590000000
4,ruq0u2ZpAMDr,419203.030000000,1.400000000
...,...,...,...
2817,iAF0NxaypyZB,6.190000000,0E-9
2818,dzTjh3Rd62QN,0.850000000,0E-9
2819,WHCTdImrIp0y,0.850000000,0E-9
2820,uzHNB354dHwV,0.850000000,0E-9


In [73]:
%%bigquery --project {project_id} --verbose

# Seller Transaction Volume Ranking Analysis

WITH seller_orders AS (

SELECT

  oi.seller_id,

  COUNT(DISTINCT o.order_id) AS total_orders,

  SUM(oi.price) AS total_sales

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  oi.seller_id

)

SELECT

  seller_id,

  total_orders,

  ROUND(total_sales,2) AS total_sales,

  RANK() OVER(
    ORDER BY total_orders DESC
  ) AS order_volume_rank

FROM seller_orders

ORDER BY
  order_volume_rank;

Executing query with job ID: 3fc47c83-c91b-4c3f-b99f-cc7a243e5f48
Query executing: 0.49s
Job ID 3fc47c83-c91b-4c3f-b99f-cc7a243e5f48 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,seller_id,total_orders,total_sales,order_volume_rank
0,RKad98cTxhSb,1624,361091.930000000,1
1,yGbPyLPc8PmT,1586,1900070.900000000,2
2,r7VxefoTVHbb,1512,2562407.580000000,3
3,TuPm19CMKvrM,1346,178643.290000000,4
4,ruq0u2ZpAMDr,1284,419203.030000000,5
...,...,...,...,...
2817,o5QOel64DoOJ,1,19.550000000,2320
2818,ApFTrBlCpCVY,1,820.050000000,2320
2819,YhEKud90h2fI,1,24.200000000,2320
2820,dlSMYy5wdITn,1,99.980000000,2320


In [74]:
%%bigquery --project {project_id} --verbose

# Seller Delivery Performance Analysis

SELECT

  oi.seller_id,

  COUNT(DISTINCT o.order_id) AS total_orders,

  COUNTIF(
    o.delivery_status = 'Late'
  ) AS late_orders,

  ROUND(

    SAFE_DIVIDE(
      COUNTIF(
        o.delivery_status = 'Late'
      ),

      COUNTIF(
        o.delivery_status IN ('Late','On Time')
      )

    ) * 100,

    2

  ) AS late_delivery_rate

FROM ecommerce.orders AS o

JOIN ecommerce.order_items AS oi

ON o.order_id = oi.order_id

WHERE
  o.delivery_status != 'Not Delivered'

GROUP BY
  oi.seller_id

HAVING
  total_orders >= 50

ORDER BY
  late_delivery_rate DESC;

Executing query with job ID: 8e62367d-43b6-4149-915b-ff6dd7e45014
Query executing: 0.51s
Job ID 8e62367d-43b6-4149-915b-ff6dd7e45014 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,seller_id,total_orders,late_orders,late_delivery_rate
0,qeM9i4J3ypy2,57,17,29.82
1,aDQK7ytnBMNf,68,19,27.94
2,FBr7bTkcVwez,62,16,25.81
3,05nIYpVrMwiz,67,17,25.37
4,c1Dw1Rwr2Qdd,52,13,25.00
...,...,...,...,...
389,tPNOMqIxsdYf,54,0,0.00
390,wnF6o59qfQ7x,53,0,0.00
391,wYEsFwaY8Gxa,50,0,0.00
392,yxfgQLIrpHFD,78,0,0.00


## Payment Analysis  

1. Which payment methods are most commonly used by customers?
2. How much does each payment method contribute to total transaction value?
3. How does installment usage relate to customer spending?
4. Which payment methods are associated with higher-value transactions?
5. How do customer payment behaviors change over time?

In [75]:
%%bigquery --project {project_id} --verbose

# Payment Method Distribution
SELECT

  payment_type,

  COUNT(DISTINCT order_id) AS total_orders,

  ROUND(
    SUM(payment_value),
    2
  ) AS total_payment_value,

  ROUND(
    AVG(payment_value),
    2
  ) AS average_payment_value

FROM ecommerce.payments

GROUP BY
  payment_type

ORDER BY
  total_orders DESC;

Executing query with job ID: e94d9b7f-49f9-40d1-8374-a690dfd46a4e
Query executing: 0.54s
Job ID e94d9b7f-49f9-40d1-8374-a690dfd46a4e successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,payment_type,total_orders,total_payment_value,average_payment_value
0,credit_card,65814,17512576.650000000,266.090000000
1,wallet,17302,4642341.690000000,268.310000000
2,voucher,4911,1523103.470000000,310.140000000
3,debit_card,1289,317363.760000000,246.210000000


In [76]:
%%bigquery --project {project_id} --verbose

# Percantage Contribution by Payment Method
WITH payment_summary AS (

SELECT

  payment_type,

  SUM(payment_value) AS total_payment_value

FROM ecommerce.payments

GROUP BY
  payment_type

)

SELECT

  payment_type,

  ROUND(
    total_payment_value,
    2
  ) AS total_payment_value,

  ROUND(

    SAFE_DIVIDE(
      total_payment_value,
      SUM(total_payment_value) OVER()
    ) * 100,

    2

  ) AS contribution_percentage

FROM payment_summary

ORDER BY
  total_payment_value DESC;

Executing query with job ID: a60f7fed-f5b5-4a2b-b6c9-a6899db41768
Query executing: 0.62s
Job ID a60f7fed-f5b5-4a2b-b6c9-a6899db41768 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,payment_type,total_payment_value,contribution_percentage
0,credit_card,17512576.650000000,72.980000000
1,wallet,4642341.690000000,19.350000000
2,voucher,1523103.470000000,6.350000000
3,debit_card,317363.760000000,1.320000000


In [77]:
%%bigquery --project {project_id} --verbose

# Payment Installments Behavior Analysis
SELECT

  payment_installments,

  COUNT(DISTINCT order_id) AS total_orders,

  ROUND(
    AVG(payment_value),
    2
  ) AS average_payment_value,

  ROUND(
    SUM(payment_value),
    2
  ) AS total_payment_value

FROM ecommerce.payments

GROUP BY
  payment_installments

ORDER BY
  payment_installments;

Executing query with job ID: 13fc3817-8ed9-4b59-ad3e-4b84402de4ef
Query executing: 0.45s
Job ID 13fc3817-8ed9-4b59-ad3e-4b84402de4ef successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,payment_installments,total_orders,average_payment_value,total_payment_value
0,0,3,600.670000000,1802.010000000
1,1,44313,269.720000000,11952066.510000000
2,2,10325,269.640000000,2784003.200000000
3,3,8907,260.180000000,2317461.390000000
4,4,6032,277.600000000,1674465.380000000
5,5,4600,272.890000000,1255279.250000000
6,6,3580,270.220000000,967398.350000000
7,7,1401,258.160000000,361677.610000000
8,8,3927,268.510000000,1054446.410000000
9,9,560,292.700000000,163911.730000000


In [78]:
%%bigquery --project {project_id} --verbose

# High Value Transaction by Payment Type
SELECT

  payment_type,

  ROUND(
    MAX(payment_value),
    2
  ) AS highest_transaction,


  ROUND(
    AVG(payment_value),
    2
  ) AS average_transaction,

  COUNT(order_id) AS total_transactions

FROM ecommerce.payments

GROUP BY
  payment_type

ORDER BY
  highest_transaction DESC;

Executing query with job ID: 47faf3fc-9764-405d-8737-3be0a53fc0d5
Query executing: 0.53s
Job ID 47faf3fc-9764-405d-8737-3be0a53fc0d5 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,payment_type,highest_transaction,average_transaction,total_transactions
0,debit_card,7274.880000000,246.210000000,1289
1,credit_card,6929.310000000,266.090000000,65814
2,voucher,6726.660000000,310.140000000,4911
3,wallet,4950.340000000,268.310000000,17302


In [79]:
%%bigquery --project {project_id} --verbose

# Monthly Payment Trend
WITH monthly_payment AS (

SELECT

  DATE_TRUNC(
    DATE(o.order_purchase_timestamp),
    MONTH
  ) AS month_date,

  FORMAT_DATE(
    '%Y-%B',
    DATE(o.order_purchase_timestamp)
  ) AS purchase_month,

  p.payment_type,

  SUM(p.payment_value) AS total_payment_value

FROM ecommerce.orders AS o

JOIN ecommerce.payments AS p

ON o.order_id = p.order_id

GROUP BY

  month_date,
  purchase_month,
  payment_type

)

SELECT

  purchase_month,

  payment_type,

  ROUND(
    total_payment_value,
    2
  ) AS total_payment_value

FROM monthly_payment

ORDER BY
  month_date,
  total_payment_value DESC;

Executing query with job ID: a53f47b6-1db5-466a-9043-b8cb94d01382
Query executing: 0.56s
Job ID a53f47b6-1db5-466a-9043-b8cb94d01382 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,purchase_month,payment_type,total_payment_value
0,2016-September,credit_card,240.190000000
1,2016-October,credit_card,55528.800000000
2,2016-October,wallet,12954.010000000
3,2016-October,voucher,3661.790000000
4,2016-October,debit_card,425.890000000
...,...,...,...
82,2018-August,credit_card,1124440.650000000
83,2018-August,wallet,240318.080000000
84,2018-August,debit_card,63785.450000000
85,2018-August,voucher,49428.130000000


In [80]:
%%bigquery --project {project_id} --verbose

# Monthly Payment Trend
WITH monthly_payment AS (

SELECT

  DATE_TRUNC(
    DATE(o.order_purchase_timestamp),
    MONTH
  ) AS month_date,

  FORMAT_DATE(
    '%Y-%B',
    DATE(o.order_purchase_timestamp)
  ) AS purchase_month,

  p.payment_type,

  SUM(p.payment_value) AS total_payment_value

FROM ecommerce.orders AS o

JOIN ecommerce.payments AS p

ON o.order_id = p.order_id

GROUP BY

  month_date,
  purchase_month,
  payment_type

)

SELECT

  purchase_month,

  payment_type,

  ROUND(
    total_payment_value,
    2
  ) AS total_payment_value

FROM monthly_payment

ORDER BY
  month_date,
  total_payment_value DESC;

Executing query with job ID: e5c57e76-e1bd-4115-8226-44cbed9e4af8
Query executing: 0.30s
Job ID e5c57e76-e1bd-4115-8226-44cbed9e4af8 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,purchase_month,payment_type,total_payment_value
0,2016-September,credit_card,240.190000000
1,2016-October,credit_card,55528.800000000
2,2016-October,wallet,12954.010000000
3,2016-October,voucher,3661.790000000
4,2016-October,debit_card,425.890000000
...,...,...,...
82,2018-August,credit_card,1124440.650000000
83,2018-August,wallet,240318.080000000
84,2018-August,debit_card,63785.450000000
85,2018-August,voucher,49428.130000000


## Advanced Analytics  

1. How are customers distributed based on spending level?
2. Do a small number of sellers contribute a large portion of revenue?
3. Which products generate the highest sales performance?

In [81]:
%%bigquery --project {project_id} --verbose

# Customer Spending Segmention
WITH customer_spending AS (

SELECT

  o.customer_id,

  SUM(p.payment_value) AS total_spending

FROM ecommerce.orders AS o

JOIN ecommerce.payments AS p

ON o.order_id = p.order_id

WHERE
  o.order_status = 'delivered'

GROUP BY
  o.customer_id

)

SELECT

CASE

  WHEN total_spending < 100 THEN 'Low Spender'

  WHEN total_spending BETWEEN 100 AND 500 THEN 'Medium Spender'

  ELSE 'High Spender'

END AS spending_segment,

COUNT(customer_id) AS total_customers

FROM customer_spending

GROUP BY
  spending_segment

ORDER BY
  total_customers DESC;

Executing query with job ID: ebea84e8-e0d8-434e-b33a-243b5579820e
Query executing: 1.04s
Job ID ebea84e8-e0d8-434e-b33a-243b5579820e successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,spending_segment,total_customers
0,Medium Spender,50377
1,Low Spender,26107
2,High Spender,10944


In [82]:
%%bigquery --project {project_id} --verbose

# Pareto Cumulative Analysis
WITH seller_sales AS (

SELECT

  seller_id,
  SUM(price) AS total_sales

FROM ecommerce.order_items

GROUP BY
  seller_id

),

seller_rank AS (

SELECT

  seller_id,
  total_sales,

  SUM(total_sales)
  OVER(
    ORDER BY total_sales DESC
  ) AS cumulative_sales,

  SUM(total_sales)
  OVER() AS overall_sales

FROM seller_sales

)

SELECT

  seller_id,
  ROUND(total_sales,2) AS total_sales,
  ROUND(
    cumulative_sales /
    overall_sales * 100,
    2
  ) AS cumulative_percentage

FROM seller_rank
ORDER BY
  total_sales DESC;


Executing query with job ID: 73dcc96c-9059-4ac0-971e-e26a2ae8f300
Query executing: 0.39s
Job ID 73dcc96c-9059-4ac0-971e-e26a2ae8f300 successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,seller_id,total_sales,cumulative_percentage
0,r7VxefoTVHbb,2569785.560000000,8.440000000
1,yGbPyLPc8PmT,1923811.420000000,14.760000000
2,K0qPVGdA91KO,666339.920000000,16.950000000
3,l1pYW6GBnPMr,480706.200000000,18.530000000
4,ruq0u2ZpAMDr,422203.730000000,19.910000000
...,...,...,...
2924,DJJi40WZb02B,6.270000000,100.000000000
2925,WHCTdImrIp0y,0.850000000,100.000000000
2926,uzHNB354dHwV,0.850000000,100.000000000
2927,u0nbR1OriQlE,0.850000000,100.000000000


In [83]:
%%bigquery --project {project_id} --verbose

# Product Performance Ranking
SELECT

  oi.product_id,
  p.product_category_name,
  COUNT(oi.order_id) AS total_orders,

  ROUND(
    SUM(oi.price),
    2
  ) AS total_sales,

  RANK()
  OVER(
    ORDER BY SUM(oi.price) DESC
  ) AS sales_rank

FROM ecommerce.order_items AS oi

LEFT JOIN ecommerce.products AS p

ON oi.product_id = p.product_id

GROUP BY
  oi.product_id,p. product_category_name

ORDER BY
  sales_rank

LIMIT 20;

Executing query with job ID: 1566b933-3779-426c-8b09-49b097b83eee
Query executing: 0.69s
Job ID 1566b933-3779-426c-8b09-49b097b83eee successfully executed


Query is running:   0%|          |

Downloading:   0%|          |

,product_id,product_category_name,total_orders,total_sales,sales_rank
0,9NwzO0Pm0fDM,toys,383,797298.940000000,1
1,SLTlrWtcYt1m,toys,321,567053.430000000,2
2,Biwi1BNtUB7l,toys,295,486503.650000000,3
3,ro08JPncYzLh,garden_tools,290,467133.530000000,4
4,ZWyg4uNWPHjJ,toys,110,318890.000000000,5
5,tPlw2QcQOvKf,garden_tools,91,153366.490000000,6
6,utVPLgd1LM3F,toys,77,146223.000000000,7
7,GvBzGCvvIC2D,toys,176,127736.000000000,8
8,5J7az1rwth4i,toys,110,117385.080000000,9
9,CY8OZ3lT9uqB,telephony,42,116979.980000000,10
